src/main/java/net/neoforged/neoforge/transfer/handlers/templates/AggregateResourceHandler.java

```diff
@@ -0,0 +1,194 @@
+    }
+
+    @Override
+    public int getCapacity(int index, T resource) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.getCapacity(index, resource);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public int getCapacity(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.getCapacity(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public boolean isValid(int index, T resource) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.isValid(index, resource);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public boolean allowsInsertion(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.allowsInsertion(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public boolean allowsExtraction(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.allowsExtraction(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public int insert(int index, T resource, int amount, TransferAction action) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                if (storage.allowsInsertion()) {
+                    return storage.insert(index, resource, amount, action);
+                } else {
+                    return 0;
+                }
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public int insert(T resource, int amount, TransferAction action) {
+        int inserted = 0;
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (storage.allowsInsertion()) {
+                inserted += storage.insert(resource, amount - inserted, action);
+            }
+            if (inserted >= amount) {
+                return inserted;
+            }
+        }
+        return inserted;
+    }
+
+    @Override
+    public int extract(int index, T resource, int amount, TransferAction action) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                if (storage.allowsExtraction()) {
```

Comment:
Do we need to check this? I feel like we should be able to just assume that the inner storage will short circuit if they aren't allowing extraction. That way if their check is for some reason not trivial, then it only has to be done once.

Reply:
Yes, it is never necessary to check `allowsXxx` before doing `xxx` (which is supposed to check it anyway). Perhaps that should be added to the javadoc.